File: bindings/nodejs/examples/exchange/1-create-wallet.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// This example creates a new database and wallet.
// Run with command:
// yarn run-example ./exchange/1-create-wallet.ts

import { Wallet, WalletOptions, CoinType, SecretManager } from '@iota/sdk';

// This example uses secrets in environment variables for simplicity which should not be done in production.
require('dotenv').config({ path: '.env' });

async function run() {
    try {
        for (const envVar of [
            'WALLET_DB_PATH',
            'NODE_URL',
            'STRONGHOLD_SNAPSHOT_PATH',
            'STRONGHOLD_PASSWORD',
            'MNEMONIC',
        ])
            if (!(envVar in process.env)) {
                throw new Error(
                    `.env ${envVar} is undefined, see .env.example`,
                );
            }

        const strongholdSecretManager = {
            stronghold: {
                snapshotPath: process.env.STRONGHOLD_SNAPSHOT_PATH,
                password: process.env.STRONGHOLD_PASSWORD,
            },
        };

        const secretManager = SecretManager.create(strongholdSecretManager);

        // A mnemonic can be generated with `Utils.generateMnemonic()`.
        // Store the mnemonic in the Stronghold snapshot, this needs to be done only the first time.
        // The mnemonic can't be retrieved from the Stronghold file, so make a backup in a secure place!
        await secretManager.storeMnemonic(process.env.MNEMONIC as string);

        const walletAddress = await secretManager.generateEd25519Addresses({
            coinType: CoinType.IOTA,
            accountIndex: 0,
            range: {
                start: 0,
                end: 1,
            },
            bech32Hrp: 'tst',
        });

        const walletOptions: WalletOptions = {
            address: walletAddress[0],
            storagePath: process.env.WALLET_DB_PATH,
            clientOptions: {
                nodes: [process.env.NODE_URL as string],
            },
            bipPath: {
                coinType: CoinType.IOTA,
            },
            secretManager: strongholdSecretManager,
        };

        const wallet = await Wallet.create(walletOptions);

        // Set syncOnlyMostBasicOutputs to true if not interested in outputs that are timelocked,
        // have a storage deposit return, expiration or are nft/account/foundry outputs.
        await wallet.setDefaultSyncOptions({ syncOnlyMostBasicOutputs: true });
    } catch (error) {
        console.error(error);
    }
}

void run().then(() => process.exit());

File: bindings/python/examples/exchange/1_create_wallet.py
# Copyright 2023 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

# This example creates a new database and wallet.

import os

from dotenv import load_dotenv
from iota_sdk import (ClientOptions, CoinType, StrongholdSecretManager,
                      SecretManager, SyncOptions, Wallet, WalletOptions, Bip44)

# This example uses secrets in environment variables for simplicity which
# should not be done in production.
load_dotenv()

for env_var in ['WALLET_DB_PATH', 'NODE_URL',
                'STRONGHOLD_SNAPSHOT_PATH', 'STRONGHOLD_PASSWORD', 'MNEMONIC']:
    if env_var not in os.environ:
        raise Exception(f'.env {env_var} is undefined, see .env.example')

client_options = ClientOptions(nodes=[os.environ.get('NODE_URL')])

secret_manager = StrongholdSecretManager(
    os.environ.get('STRONGHOLD_SNAPSHOT_PATH'), os.environ['STRONGHOLD_PASSWORD'])

# Store the mnemonic in the Stronghold snapshot, this only needs to be
# done once.
SecretManager(secret_manager).store_mnemonic(os.environ['MNEMONIC'])

bip_path = Bip44(
    coin_type=CoinType.SHIMMER
)
wallet_options = WalletOptions(
    None,
    None,
    bip_path,
    client_options,
    secret_manager,
    os.environ.get('WALLET_DB_PATH'))
wallet = Wallet(wallet_options)

# Set sync_only_most_basic_outputs to True if not interested in outputs that are timelocked,
# have a storage deposit return, expiration or are nft/account/foundry outputs.
wallet.set_default_sync_options(
    SyncOptions(sync_only_most_basic_outputs=True))

# Update the wallet to the latest state
balance = wallet.sync()
print('Generated new wallet')
